compiler/test/stdlib/json.test.gr

```diff
@@ -0,0 +1,557 @@
+import * from "json"
+import Result from "result"
+import List from "list"
+
+// Parsing
+
+// Simple values of each type
+assert parse("null") == Ok(JSONNull)
+assert parse("true") == Ok(JSONBoolean(true))
+assert parse("false") == Ok(JSONBoolean(false))
+assert parse("\"\"") == Ok(JSONString(""))
+assert parse("0") == Ok(JSONNumber(0))
+assert parse("[]") == Ok(JSONArray([]))
+assert parse("{}") == Ok(JSONObject([]))
+
+// Various strings. Escapes, emojis etc.
+assert parse("\"ASCII Hello world!\"") == Ok(JSONString("ASCII Hello world!"))
+
+assert parse("\"Unicode こんにちは世界!\"") ==
+Ok(JSONString("Unicode こんにちは世界!"))
+assert parse("\"A \\\"quoted\\\" string\"") ==
+Ok(JSONString("A \"quoted\" string"))
+assert parse("\"\\uD801\\uDC37\"") == Ok(JSONString("𐐷"))
+assert parse("\"🤘🏻\"") == Ok(JSONString("🤘🏻"))
+assert parse("\"🤘🏻\"") == Ok(JSONString("🤘🏻"))
+assert parse("\"🏴󠁧󠁢󠁷󠁬󠁳󠁿\"") ==
+Ok(JSONString("🏴󠁧󠁢󠁷󠁬󠁳󠁿"))
+assert parse(
+  "\"\\u6000 \\ud800\\udc82 \\ud83e\\udd18\\ud83c\\udffb \\ud83c\\udff4\\udb40\\udc67\\udb40\\udc62\\udb40\\udc77\\udb40\\udc6c\\udb40\\udc73\\udb40\\udc7f\""
+) ==
+Ok(JSONString("怀 𐂂 🤘🏻 🏴󠁧󠁢󠁷󠁬󠁳󠁿"))
+
+// Different number types and values
+assert parse("-10") == Ok(JSONNumber(-10))
+assert parse("-2") == Ok(JSONNumber(-2))
+assert parse("-1") == Ok(JSONNumber(-1))
+assert parse("1") == Ok(JSONNumber(1))
+assert parse("2") == Ok(JSONNumber(2))
+assert parse("10") == Ok(JSONNumber(10))
+assert parse("100") == Ok(JSONNumber(100))
+assert parse("1000") == Ok(JSONNumber(1000))
+assert parse("0.0") == Ok(JSONNumber(0.0))
+assert parse("0.1") == Ok(JSONNumber(0.1))
+assert parse("0.123") == Ok(JSONNumber(0.123))
+assert parse("0.9") == Ok(JSONNumber(0.9))
+assert parse("1.123") == Ok(JSONNumber(1.123))
+assert parse("0e0") == Ok(JSONNumber(0.0))
+assert parse("1e0") == Ok(JSONNumber(1.0))
+assert parse("1e1") == Ok(JSONNumber(10.0))
+assert parse("1E1") == Ok(JSONNumber(10.0))
+assert parse("1e2") == Ok(JSONNumber(100.0))
+assert parse("1e3") == Ok(JSONNumber(1000.0))
+assert parse("-1e2") == Ok(JSONNumber(-100.0))
+assert parse("1e-1") == Ok(JSONNumber(0.1))
+assert parse("1.23e-4") == Ok(JSONNumber(0.000123))
+
+// FIXME
+// "-0" currently parses to 0, but should it?
+// assert parse("-0") == Ok(JSONNumber(?))
+// Note that in Grain these are true:
+// assert 0 == 0
+// assert 0.0 == -0.0
+// assert 0 == -0.0
+// And printing constant "-0" results in "0"
```

Comment:
I'm fine with `-0` parses to `0`, but that's just my take

Reply:
It's one of these arbitrary decisions to make. While whether negative zero exists is more of the domain of philosophy and mathematics, it does exist in IEEE 754-2008. I'm in favor of parsing into -0.0, but only based on the argument that parsing and reprinting would otherwise technically result loss of information.

Reply:
I agree with cician though I also think grain should parse -0 as -0 because if you were Todo if (parsedjsonint == -0) would be false which can be confusing for some users.

Reply:
Just to add some context: Grain leaves out the minus sign from a Number constant, but allows creating an instance of Float64 with value -0.0 and cast it into a Number.